### ISSUE TITLE
[5.0.0] Combination of XElement & relative swagger endpoint causes recursion and errors in swagger ui

### ISSUE THREAD

**Issue by Darkov3** [domaindrivendev/Swashbuckle.AspNetCore#1492]
When using a relative path(such as "./swagger/v1/swagger.json") in the SwaggerEndpoint for the UI with a circular object(such as an XElement), this causes strange errors and recursion in the network tab. The same configuration worked with .NET Core 2 and v4 of Swashbuckle.AspNetCore.
This is done using NewtonsoftJson, because System.Text.Json has an issue with circular objects too.

**Problem:**
![Capture](https://user-images.githubusercontent.com/5776181/72738936-4e75a280-3bab-11ea-9af4-26dec7814502.PNG)

![Capture](https://user-images.githubusercontent.com/5776181/72739031-87ae1280-3bab-11ea-9b92-75cd0b85c0c6.PNG)

**Reproduction:**
Nugets:
![Capture](https://user-images.githubusercontent.com/5776181/72739314-11f67680-3bac-11ea-86b4-7c87114df58f.PNG)

Startup.cs:
```c#
public void ConfigureServices(IServiceCollection services)
{
    services.AddControllers()
        .AddNewtonsoftJson();
    services.AddSwaggerGen(c =>
    {
        c.SwaggerDoc("v1", new OpenApiInfo { Title = "Test API", Version = "v1" });
    });
    services.AddSwaggerGenNewtonsoftSupport();
}

public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{
    app.UseSwagger();

    app.UseSwaggerUI(c =>
    {
        c.SwaggerEndpoint("./swagger/v1/swagger.json", "Test API V1");
        c.RoutePrefix = "";
    });

    if (env.IsDevelopment())
    {
        app.UseDeveloperExceptionPage();
    }

    app.UseHttpsRedirection();

    app.UseRouting();

    app.UseAuthorization();

    app.UseEndpoints(endpoints =>
    {
        endpoints.MapControllers();
    });
}
```

TestController.cs:
```c#
[ApiController]
[Route("[controller]")]
public class TestController : ControllerBase
{
    private readonly ILogger<TestController> _logger;

    public TestController(ILogger<TestController> logger)
    {
        _logger = logger;
    }

    [HttpGet]
    public ActionResult<XElement> Test()
    {
        var rng = new Random();
        return XElement.Parse(@"<test>" + rng.Next(0, 101) + @"</test>");
    }
}
```

Run the test service and go to https://localhost:5001/index.html and expand the /Test method to trigger the error.

---
**Conclusion:**
We are forced to do some magical things to get around this issue, any workarounds are welcome too.

**Comment by mikevanoo** [comment]
I'm hitting the same issue. #903 recommends using various forms of relative url but they all hit this issue.

**Comment by Darkov3** [comment]
The workarounds we tried all failed.
Since the error does not seem to actually break anything we added this line, for production, to hide the errors instead:
```c#
c.HeadContent += @"<style>.errors-wrapper { display: none !IMPORTANT; }</style>";
```

**Comment by mikevanoo** [comment]
Yikes! 

It's not just the errors on screen. It slows everything down as well whilst it fires off dozens of requests for swagger.json at invalid urls.

The only thing that works is a virtual path, i.e. /swagger/v1.0/swagger.json

However, we deploy to different sub-paths in different environments so we'll have to pass a "BasePath" config value in and use that, e.g. /<BasePath>/swagger/v1.0/swagger.json

**Comment by mikevanoo** [comment]
... "we'll have to pass a "BasePath" config value in and use that"... 

... because we haven't found a way to determine the base path of an Asp.Net Core site during Startup::Configure(). It seems you always need a HttpContext or be inside a Controller.

**Comment by domaindrivendev** [comment]
Why the "./" prefix? The following setup uses a relative path (i.e. relative to where the UI is served at) and works just fine for me, regardless of where the UI is ultimately served at:

```csharp
app.UseSwagger();

app.UseSwaggerUI(c =>
{
    c.RoutePrefix = ""; // serve the UI at root
    c.SwaggerEndpoint("swagger/v1/swagger.json", "V1 Docs");
});
```

**Comment by Darkov3** [comment]
> Why the "./" prefix? The following setup uses a relative path (i.e. relative to where the UI is served at) and works just fine for me, regardless of where the UI is ultimately served at:
> 
> ```cs
> app.UseSwagger();
> 
> app.UseSwaggerUI(c =>
> {
>     c.RoutePrefix = ""; // serve the UI at root
>     c.SwaggerEndpoint("swagger/v1/swagger.json", "V1 Docs");
> });
> ```

I have tested your proposed variant and even the ../ relative path variant and I still get the same erroneous result.
I tested your suggestion in the example in the OP here and I get the same erroneous result. If you want I can upload the test solution and if you do not get the same result with it, then the problem must be elsewhere.

**Comment by mikevanoo** [comment]
I agree with @Darkov3 . Its a combination of a relative SwaggerEndpoint path (any way of doing it) and some kind of circular reference in the model(s).

**Comment by domaindrivendev** [comment]
@Darkov3 - as I can't seem to repro locally, I think it would be best if you create a minimal project from scratch that repro's the issue, and then push that up to github where I can pull it down and hopefully get a repro to work with

**Comment by Darkov3** [comment]
@domaindrivendev This is the project from the OP here, I've tested it on 2 computers:
https://github.com/Darkov3/swagger-ui-v5-test

**Comment by domaindrivendev** [comment]
OK - so the root cause is that Swashbuckle is describing the response to be what it "thinks" an `XElement` instance would look like when serialized to JSON. It's treating it as a plain old DTO object, where all public properties and fields pretty much map 1:1 to JSON object properties. Of course, this looks nothing like the _actual_ JSON structure that's returned by the service. This is because Newtonsoft treats `XElement` differently (like a unique snowflake). Rather than basing the JSON on public properties/fields of the class itself, it takes the XML string captured in a given instance, and then converts that to JSON.

This has two repurcusions in the context of Swashbuckle.
1) It doesn't currently support the nuanced Newtonsoft behavior around `XElement` (this can be added but with a limitation - see next point)
2) When describing responses that are based on `XElement`, the best Swashbuckle can do is describe it as an "open schema". That is, it can't provide any information about the structure of the JSON response. This is because the structure is determined at run-time (i.e. when a given `XElement` is instantiated and assigned values) and therefore can't be inferred by Swashbuckle from the code alone.

I'll work to add some level of support for the `XElement` type (with the limitation mentioned above) but in the meantime, you could support it in the same way with a custom type mapping:

```csharp
services.AddSwaggerGen(c =>
{
    c.SwaggerDoc("v1", new OpenApiInfo { Title = "Test API", Version = "v1" });

    c.MapType<XElement>(() => new OpenApiSchema());
});
```

**Comment by domaindrivendev** [comment]
OK interesting ... it seems the UI not handling the ugly, complex schema that's incorrectly being generated for `XElement` is somehow only happening when relative paths are in use. This needs further investigation on my side. I have a feeling it might be related to self-referencing types. I'll try get a tighter repro for that.

Either way, you can probably avoid this issue by doing what I suggested, which would eradicate the problematic schema entirely. I imagine having an "open schema" would be preferable anyway to the complexy schema that actually looks nothing like the response that's ultimately returned.

**Comment by Darkov3** [comment]
Thank you, this does solve my particular problem. The generated XElement Schema was not useful anyway and the open api scheme is not that useful too, but it does not generate the errors, which is what the real issue was for me.

**Comment by mikevanoo** [comment]
@domaindrivendev Thanks for your time on this. And I agree, definitively something to do with the combination of self-referencing types and relative paths.

Sadly, for us, our model contains self-referencing types but they are genuine and well-defined at compile-time, so not like XElement. Any further investigation you can do would be much appreciated.

**Comment by inbarbarkai** [comment]
It happened to me when I have entities that are partially documented.
For example, I'm using IdentityServer and one of my endpoints is returning `IdentityServer4.Models.Client[]`.
The `IdentityServer4.Models.Client` contains reference to `System.Security.Claims.Claim` and `System.Security.Claims.ClaimsIdentity`.
I don't add the XML documentation for the `System.Security.Claims.dll` library.
The result is multiple requests to the server to get the `swagger.json` file but the URL is wrong.
Keeps adding v1: `/swagger/v1/swagger.json`...`/swagger/v1/v1/swagger.json` and so on.

**Comment by remcolam** [comment]
If I can weigh in, I'm also running into this now, the combination of a relative path and a circular reference:

```
				app.UseSwaggerUI(options =>
				{
					options.SwaggerEndpoint($"../swagger/{API.Constants.Swagger.DocumentationGroups.Admin}/swagger.json", API.Constants.Swagger.DocumentationGroups.Admin.Capitalize());
```
The relative path is needed, basepath is variable depending on where we are hosting , sometimes basepath is the root, sometimes a folder, and different names folders at that

I've done some testing with relative paths and the route-prefix, but there is no working combination. as soon as a circular reference is in the model, it will generate errors, unless the SwaggerEndpoint is started with a forward slash, but that means you always need to include that variable basepath in you swagger endpoint configuration

**Comment by domaindrivendev** [comment]
I believe this is essentially a dup of #1545 so going to close this so it can be tracked in one place. Long story short though, it's due to a bug in the [swagger-ui library](https://github.com/swagger-api/swagger-ui), which is independent of Swashbuckle.

I went ahead and created the following [issue](https://github.com/swagger-api/swagger-ui/issues/5945) in the swagger-ui repo. But, unfortunately no response yet 😞

I'm investigating potential workarounds that can be applied on the SB side, but ideally it would be fixed in the upstream project.

**Comment by domaindrivendev** [comment]
In lieu of an upstream fix, I've just checked [a SB change](98a7dfe0f7ca2aa66c6daa029c3987490bd1eb20) to workaround the issue by expanding "page-relative" URLs to a fully-qualified URLs on the client-side before passing to the `SwaggerUI` javascript component:

You can try it out by pulling the [latest SB preview build](https://myget.org/feed/domaindrivendev/package/nuget/Swashbuckle.AspNetCore) from myget.org. Let me know how it goes?

**Comment by freeranger** [comment]
@domaindrivendev I had exactly this problem too and your workaround does the trick.  Thanks
